### ISSUE TITLE
put-file show work with same HTTP authentication types curl/wget do

### ISSUE THREAD

**Issue by pappasilenus** [pachyderm/pachyderm#3435]
**What is the goal / desired outcome?**

As a data scientist, when using put-file to put data into pachyderm from urls, I would like to embed HTTP authentication tokens in the request so that I can use backends that require authentication to get data.

**If there is a way to accomplish this today via workaround, what does that require?**
 I can use curl or wget to stage the data to my machine first.  This creates a security and privacy risk.  I would rather the data were transferred directly into pachyderm from the storage url.

**(Optional) What is your proposal for a feature to solve this?**

Good

Allow put-file to use the same authentication syntax as curl or wget (-u or --user <username: password>) and work with standard HTTP authentication mechanisms.

Better

Allow for the addition of headers to the outgoing HTTP requests for custom authentication schemes, as in curl's "-H" flag (as in curl -H "X-Auth-User: superman" -H "X-Auth-Key: superman")

Best

Prompt for authentication to the end user if the response from the resource demands it, caching credentials as needed.

**Comment by jdoliner** [comment]
@gabrielgrant that's correct, I think #2868 is the simplest thing we could do that would satisfy the use case.  However there's still probably some need for http authentication. I'm not sure quite how deep we want to go on this functionality. Recreating all of curls auth flags is probably overdoing it a bit, at a certain point it probably makes sense for people to just run:

```
curl ... | pachctl put-file ...
```

From a pod within the cluster, that way you have complete freedom to do whatever you want and the performance should be pretty close to equivalent.

**Comment by jdoliner** [comment]
@gabrielgrant exactly, it would work wherever those style of urls are used. This is in reference a comment on #2868 for those wondering where `s3api` came from.

**Comment by ysimonson** [comment]
I'm a fan of just piping from curl for any type of authentication. It feels more unix-ey, provides better separation of concerns, doesn't carry a massive performance tax, and doesn't have a security/privacy risk since the contents aren't persisted to disk.
